queue/driver_mongo.go

```diff
@@ -1400,14 +1268,146 @@ type dispatchAttemptInfo struct {
+	if err != nil {
+		return nil, info, errors.Wrap(err, "dispatching stale in progress job")
+	}
+
+	if job != nil || d.opts.SampleSize > 0 && dispatchInfo.total >= d.opts.SampleSize {
+		return job, info, nil
+	}
+
+	pendingSampleSize := 0
+	if d.opts.SampleSize > 0 {
+		pendingSampleSize = d.opts.SampleSize - dispatchInfo.total
+	}
+	job, dispatchInfo, err = d.tryDispatchJob(ctx, d.getNextPendingQuery(now), pendingSampleSize, startAt)
+	info.misses += dispatchInfo.misses
+	info.skips += dispatchInfo.skips
+	if err != nil {
+		if err != nil {
+			return nil, info, errors.Wrap(err, "dispatching stale in progress job")
+		}
+	}
+
+	return job, info, nil
+}
+
+// getNextStaleInProgressQuery returns the query for the next available stale in-progress jobs.
+func (d *mongoDriver) getNextStaleInProgressQuery(now time.Time) bson.M {
+	staleBefore := now.Add(-d.LockTimeout())
+	qd := d.getInProgQuery(bson.M{"status.mod_ts": bson.M{"$lte": staleBefore}})
+	d.modifyQueryForGroup(qd)
+	qd = d.modifyQueryForTimeLimits(qd, now)
+
+	return qd
+}
+
+// getNextPendingQuery returns the query for the next available pending jobs.
+func (d *mongoDriver) getNextPendingQuery(now time.Time) bson.M {
+	qd := d.getPendingQuery(bson.M{})
+	d.modifyQueryForGroup(qd)
+	qd = d.modifyQueryForTimeLimits(qd, now)
+
+	return qd
 }
 
-// tryDispatchJob takes an iterator over the candidate Amboy jobs and attempts
+// modifyQueryForTimeLimits returns a new query that adds a check to the given query for jobs' time limits.
+// If neither CheckWaitUntil nor CheckDispatchBy are set the original query is returned.
+func (d *mongoDriver) modifyQueryForTimeLimits(q bson.M, now time.Time) bson.M {
+	var timeLimits []bson.M
+	if d.opts.CheckWaitUntil {
+		checkWaitUntil := bson.M{"$or": []bson.M{
+			{"time_info.wait_until": bson.M{"$lte": now}},
+			{"time_info.wait_until": bson.M{"$exists": false}},
+		}}
+		timeLimits = append(timeLimits, checkWaitUntil)
+	}
+
+	if d.opts.CheckDispatchBy {
+		checkDispatchBy := bson.M{"$or": []bson.M{
+			{"time_info.dispatch_by": bson.M{"$gt": now}},
+			{"time_info.dispatch_by": bson.M{"$exists": false}},
+		}}
+		timeLimits = append(timeLimits, checkDispatchBy)
+	}
+
+	if len(timeLimits) > 0 {
+		q = bson.M{"$and": append(timeLimits, q)}
+	}
+
+	return q
+}
+
+// tryDispatchJob attempts to dispatch a job that matches the query. If sampleSize is greater than
+// zero only that number of jobs are considered and the jobs are randomly shuffled.
+// If dispatching succeeds, the successfully dispatched job is returned.
+func (d *mongoDriver) tryDispatchJob(ctx context.Context, query bson.M, sampleSize int, startAt time.Time) (amboy.Job, dispatchAttemptInfo, error) {
+	iter, err := d.getNextCursor(ctx, query, sampleSize)
+	if err != nil {
+		return nil, dispatchAttemptInfo{}, errors.Wrap(err, "getting next job iterator")
+	}
+
+	job, dispatchInfo := d.tryDispatchFromCursor(ctx, iter, startAt)
+
+	if err = iter.Err(); err != nil {
```

Comment:
If this errors, we have to still ensure we release the job from the dispatcher if it's non-nil.

Reply:
returned the job